requests/adapters.py

```diff
@@ -272,13 +272,10 @@ def request_url(self, request, proxies):
         :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.
 
         :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
-        :param proxies: A dictionary of schemes to proxy URLs.
+        :param proxies: A dictionary of schemes or schemes and hosts to proxy URLs.
         """
-        proxies = proxies or {}
-        scheme = urlparse(request.url).scheme
-        proxy = proxies.get(scheme)
-
-        if proxy and scheme != 'https':
+        proxy = select_proxy(request.url, proxies)
+        if proxy and not request.url.lower().startswith('https'):
```

Comment:
FYI, it looks like there is an order of magnitude difference on my machine between using startswith and using urlparse, from about 300 ns for startswith, to 3 microseconds for urlparse.